accepted/install-locations.md

```diff
@@ -0,0 +1,227 @@
+.NET Core is "installed" by simply creating the right layout on the disk with the right files, there's no system registration mechanism attached to it (typically called x-copy deployment).
+
+This can be used for
+* Project-local .NET Core SDK/runtime (for example most dotnet repos use this)
+* Experimentation
+
+As such it's actually desirable that creating such installation doesn't affect anything which didn't opt into using it.
+
+*This type of install is supported by .NET Core 2.2 and earlier.*
+
+### Self-contained apps
+Technically speaking self-contained apps also "install" a private copy of the runtime, but they don't use the same layout and such installation is only usable by the app itself. So for the purposes of this document, self-contained apps are largely out of scope.
+
+
+## Components
+These are the components of .NET Core which must be aware of the install location:
+* The host (muxer (`dotnet`), `apphost`, `comhost`, ...) - the host needs to be able to find the location to search for
+  * Latest version of `hostfxr` to use.
+  * Shared frameworks
+  * SDKs
+* Installer - the installer needs to put the files into the right layout and register the install location with the system configuration (if applicable)
+* Applications
+  * If the application is using the muxer (`dotnet`) it needs to be able to find it. This applies to tools like MSBuild or VS as well.
+  * Certain tools (installers, development tools) may need to determine if .NET Core is installed and which versions are available.
+
+
+## Servicing considerations
+The algorithm which determines where to search, is to some degree contained in all the components mentioned above. Changes to the algorithm (for example introduction of a new default location or new install type) need to be applied to all of these components.
+
+Unfortunately some of these components are hard to service:
+* The muxer relies on the caller to locate it, then it's shared by all versions installed in any given location. Updating it is troublesome as it typically require a system reboot (may be in use by running applications).
+* The `hostfxr` is relatively easy to service since the installer places a newer version side-by-side and relies on the host to pick the latest available version. The complexity comes from the fact that it has to work with all versions of the runtime/sdk.
+* App-contained hosts (`apphost`, `comhost`, `ijwhost`, `nethost`) these are libraries which are shipped with the apps and thus fully owned/serviced by the app. They need to know how to find install locations to locate shared frameworks.
+
+The biggest problem is with the app-contained hosts which we can't service in any way. So introducing new install types or changing the default location can be breaking for apps using old hosts.
+
+The problem already exists since the `apphost` has been shipped in .NET Core 2.1 with support for framework dependent apps. Fortunately this feature is not widely used yet.
```

Comment:
It is widely used in global tools. 

It's not how many tools there are, but how many people are using those tools. 

But I don't see any breaking changes proposed.

Reply:
I added the global tools case.

Please note that the problem is not breaking changes in the `apphost`, but rather the effect of introducing new install locations in the future.
If I have an `apphost` 2.2 app on a clean machine which only installs into a custom location (registered in registery) as can happen in the 3.0 case.  Even if that app has "roll forward on major" and should run on the 3.0 just fine, it won't, because it won't find it.

Since in 3.0 majority of apps will have `apphost`, if we in 4.0 we introduce a new install location this problem will be greatly exagerated. The urgency comes from this fact. It would be better if we can agree on set of install locations now, which we think should work well in the future. Then 3.0 `apphost` should work going forward.

I know that we can't predict the future, but this is an attempt to mitigate the risk of the above mentioned problem in the future. (And also to write down the existing and desired behavior somewhere, since we don't seem to have that anywhere really).